controllers/machinehealthcheck_controller.go

```diff
@@ -0,0 +1,69 @@
+package controllers
+
+import (
+	"context"
+
+	"github.com/go-logr/logr"
+	"github.com/medik8s/node-healthcheck-operator/controllers/cluster"
+	"github.com/medik8s/node-healthcheck-operator/controllers/mhc"
+	"github.com/medik8s/node-healthcheck-operator/controllers/utils"
+	"github.com/openshift/api/machine/v1beta1"
+	v1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/client-go/tools/record"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+// MachineHealthCheckReconciler reconciles a MachineHealthCheck object
+type MachineHealthCheckReconciler struct {
+	client.Client
+	Log                         logr.Logger
+	Scheme                      *runtime.Scheme
+	Recorder                    record.EventRecorder
+	ClusterUpgradeStatusChecker cluster.UpgradeChecker
+	MHCChecker                  mhc.Checker
+}
+
+// +kubebuilder:rbac:groups=core,resources=nodes,verbs=get;list;watch
+// +kubebuilder:rbac:groups=config.openshift.io,resources=clusterversions,verbs=get;list;watch
+// +kubebuilder:rbac:groups=machine.openshift.io,resources=machinehealthchecks,verbs=get;list;watch;update;patch
+// +kubebuilder:rbac:groups=machine.openshift.io,resources=machinehealthchecks/status,verbs=get;update;patch
+// +kubebuilder:rbac:groups=machine.openshift.io,resources=machinehealthchecks/finalizers,verbs=update
+
+// Reconcile is part of the main kubernetes reconciliation loop which aims to
+// move the current state of the cluster closer to the desired state.
+func (r *MachineHealthCheckReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	log := r.Log.WithValues("MachineHealthCheck", req.NamespacedName)
+
+	// update MHCChecker status
+	r.MHCChecker.UpdateStatus()
+
```

Comment:
IIUC  this always run (even where MHC is not the detection method),
Therefore I'd expect here r.MHCChecker.NeedDisableNHC() check.
In case there is no MHC we can skip the rest of reconcile.

Reply:
MHCChecker.NeedDisableNHC() is used by the NHC reconciler for checking if NHC conflicts with MHC. Before this PR, every call of NeedDisableNHC() (so every NHC reconclile) was listing MHCs for making that decision.
With this PR, we only list MHCs and update checker status when a MHC was changed. That's the only thing the MHC reconciler does at the moment, beside some boilerplate code + logging, which prepares the next step of dealing with MHCs.

Reply:
>With this PR, we only list MHCs and update checker status when a MHC was changed.

IIUC this infect means that this reconcile will only be activated on "relevant" cases which makes a lot of sense.

However I'm don't completely understand how do we achieve this.
For example IIUC `// +kubebuilder:rbac:groups=core,resources=nodes,verbs=get;list;watch` this means that the MHC reconcile will be triggered by a change to a node which can occur even when NHC is the default.

Reply:
> IIUC this infect means that this reconcile will only be activated on "relevant" cases which makes a lot of sense.

I do not understand, what do you mean with "infect" and "relevant cases"?

> // +kubebuilder:rbac:groups=core,resources=nodes,verbs=get;list;watch

This just for generating RBAC roles.
What you mean is here:
```// SetupWithManager sets up the controller with the Manager.
func (r *MachineHealthCheckReconciler) SetupWithManager(mgr ctrl.Manager) error {
	return ctrl.NewControllerManagedBy(mgr).
		==> For(&v1beta1.MachineHealthCheck{}).
		==> Watches(&source.Kind{Type: &v1.Node{}}, handler.EnqueueRequestsFromMapFunc(utils.MHCByNodeMapperFunc(mgr.GetClient(), mgr.GetLogger()))).
		    Complete(r)
}
```
But you are right, watching nodes is not needed for this PR 👍🏼 It is is a "leftover" for the next step.

Reply:
>  watching nodes is not needed for this PR

removed (and some more unneeded stuff)

Reply:
Might not have explained myself properly.
In a nutshell what I meant, is that because nodes are watched both (NHC & MHC) reconcile will be triggered upon change which IIUC was not the goal.